### DISCUSSION TITLE
Chapter 05. 관계 데이터 모델

### DISCUSSION THREAD

**Discussion by crtEvent** [CodeSquad-2023-BE-Study/DB-Study#8] [완료]
책을 읽으면서 궁금한 점을 남겨주세요.

**Comment by crtEvent** [comment]
책 124(PDF 122p)
> 릴레이션은 파일 관리 시스템의 파일, 속성은 해당 파일의 필드field에 대응하는 개념
> 투플은 파일 관리 시스템 관점에서 해당 파일의 레코드record에 대응하는 개념이다.

## 파일 시스템과 파일 관리 시스템은 다른 것인가?
### 파일 시스템
- 컴퓨터에서 파일이나 자료를 쉽게 발견할 수 있도록, 유지 및 관리하는 방법
- 파일(데이터의 모임)을 저장 장치에 저장하고 사용하기 위한 일종의 규칙이나 체계

### 파일 관리 시스템
- 파일을 관리하기 위한 소프트웨어 시스템이나 응용 프로그램
- 파일의 생성, 복사, 이동, 삭제 등과 같은 파일 조작을 수행하는 기능을 제공
- e.g. 운영 체제의 파일 탐색기

**Comment by crtEvent** [comment]
## 필드, 레코드는 파일 관리 시스템에 쓰이는 용어인가?
- 그냥 관계형 DB 테이블에서도 필드와 레코드라는 이름을 쓰는 걸 봐서 사람들이 헷갈려서 쓰는 건지 원래 같이 쓰는 건지?

**Reply by leegyeongwhan** [reply]
ㅋㅋㅋㅋㅋ 저도 처음 책을 봤을 때 그용어가 그용어 같고 헷갈리네요

**Reply by jaea-kim** [reply]
필드와 레코드도 데이터베이스 관리 시스템에서도 사용되고 파일 관리 시스템에서도 사용되지만 개념은 다른 거로 알고 있습니다

**Reply by new-pow** [reply]
저도 궁금해서 관련 자료를 찾아보았는데 정확하지는 않으나 이 게시물이 그나마 간략하게 잘 설명해준 것 같아 공유합니다.

[링크](https://chaengstory.tistory.com/32)

**Comment by crtEvent** [comment]
책 138p ~ 139p
> 외래키는 자신이 참조하는 릴레이션에 기본키 값으로 존재하는 값, 즉 참조 가능한 값만 가져야 한다.

> 외래키가 널 값인 경우는 참조 무결성 제약조건을 위반한 것인가?
> 그렇지 않다. 주문고객 속성 값이 널이라는 것은 주문한 고객이 누구인지 모를 뿐, 고객 릴레이션에 존재하지 않는 고객이 주문한 것으로 판단하기는 어렵기 때문이다.

> 참조 무결성 제약 조건을 만족 시키려면 외래키가 참조 가능한 값만 가져야 하지만, 널 값을 가진다고 해서 참조 무결성 제약조건을 위반한 것은 아니다.

- 참조하는 릴레이션에 기본키 값으로 존재하는 값만 가져야 한다는 제약 조건을 걸어 놓고 기본키가 절 대 가질 수 없는 Null값은 허용한다는게 납득이 가지 않네요.

- 주문고객 속성 값이 널이라는 것은 주문한 고객이 누구인지 모를 뿐 = 외래키의 속성 값이 널이라는 것은 참조하는 릴레이션의 기본값 속성이 무엇인지 모른다.
- 모른다는 건 기본키에 없는 값을 가질 수도 있다는 거 아닌가??? 슈뢰딩거의 고양이 같은 느낌
- 외래키가 두 릴레이션을 연관 시키려고 있는 건데 Null이면 어떻게 연관되었다고 말할 수 있는 걸까?
- 어떤 투플은 연결 되어 있고 어떤 투플은 연결 되어있지 않고 이걸 1:n, n:m 이런 관계로 정의할 수 있는 걸까?

**Reply by new-pow** [reply]
ㅋㅋㅋㅋ 질문 읽으면서 정신이 혼미해져갑니다.
제가 이해한바로는
1. '외래키가 두 릴레이션을 연관시킨다'는 것은 스키마에서 정의되지 않을까요?
2. null 은 어떤 값이랑 연관되었는지 모른다는 것은 오히려 현실 세계랑 접목시키면 이해가 쉬울 것 같습니다.
    - 회원 릴레이션에 추천인 속성은 회원 릴레이션을 자기 참조하고 있지만 추천인이 없을 경우 null 일 수 있겠죠. 하지만 이것이 곧 릴레이션간의 관계가 없다기 보다는 해당 튜플이 추천인이 없다는 것을 가르킵니다. 이것이 데이터 무결성을 헤치지 않는다는 것이 책의 답변이고요

**Comment by leegyeongwhan** [comment]
외래키가 널 값인 경우는 참조 무결성 제약조건

그전부터 외래키가 null인경우도 허용 되는 것으로는 알고있었는데 사용할 때 주의 해야 한다는 것을 처음 알게 되었네요 특히 조회할때
처리를 어떻게 해주냐에 따라 결과가 달라지는군요 null의 의미는 db에서  제가 기존에 알고 있던 null의 의미와는 다른 것 같습니다.

https://www.youtube.com/watch?v=y_7rOoOodCY

**Comment by new-pow** [comment]
그래서 릴레이션이 무엇인지....??

**Reply by jinny-l** [reply]
ㅋㅋ약간 모호하긴한데 제가 느끼기엔 테이블이랑 비슷한 개념인 것 같아요

**Reply by jaea-kim** [reply]
테이블이라고 봐도 되지 않을까요..
제가 찾아본 바로는 
- 릴레이션 : 관계형 데이터베이스에서 데이터를 구조화하고 저장하는 데 사용되는 개념
- 테이블 : 구체적인 데이터베이스 관리 시스템의 용어라고 봐야한다고 봤습니다. 일종의 테이블이 릴레이션의 구현체라고 본다고 하네요

**Comment by jinny-l** [comment]
p125

사실 4장에서도 널값 관련 내용이 있었는데 5장에서도 또나와서 논쟁 거리 하나 던져봅니다.

Entity에서 Wrapper Type을 쓴다면 왜 쓰는지?
Entity에서 Primitive Type을 쓴다면 왜 쓰는지?
각각의 장단점?

**Reply by crtEvent** [reply]
Wrapper Type 장점: Null 값이 들어와도 NullPointException 안 뜬다.

**Reply by crtEvent** [reply]
Primitive Type 단점: user id값은 0, null인 경우를 어떻게 확인함??

**Reply by new-pow** [reply]
- 참고링크 : [Entity의 field type은 무엇이 적합할까](https://hyune-c.tistory.com/44)

**Reply by new-pow** [reply]
- 시퀀스전략 : https://junhyunny.github.io/spring-boot/jpa/jpa-sequece-key-generate-strategy/

**Comment by jinny-l** [comment]
키의 종류는 대충 어떤 종류가 있는지랑 개념은 다 아는데,
각각의 용어와 개념이 매칭되지 않을 때가 많아서
달달 외워야 할지 아니면 개념정도만 알고 넘어갈지 볼때마다 매번 고민이네요.
